src/script/script.cpp

```diff
@@ -227,3 +238,29 @@ bool CScript::IsPushOnly() const
     }
     return true;
 }
+
+int CScript::Type() const
+{
+    if (this->IsPayToPublicKeyHash())
+        return 1;
+    if (this->IsPayToScriptHash())
+        return 2;
+    // We don't know this script
+    return 0;
+}
+
+// This method should only be called if Type() returned nonzero
+vector<unsigned char> CScript::AddressHash() const
+{
+    int start = 0;
+    if (this->IsPayToScriptHash()) {
+        start = 2;
+    } else if (this->IsPayToPublicKeyHash()) {
+        start = 3;
+    } else {
+        assert(false);
+    }
+
+    vector<unsigned char> hashBytes(this->begin()+start, this->begin()+start+20);
```

Comment:
You could change this method to return a `uint160`. It looks like everywhere it is called we do something like `uint160 const addrHash = uint160(out.scriptPubKey.AddressHash(scriptType));` so this could be inlined in to this method.

Reply:
Also, `CScript::AddressHash(int scriptType)` could be modified so that we do not pass in the scriptType. Instead of 
```
    switch (scriptType) {
    case 1: // P2PKH
    ...
```
we could do 
```
    if (this->IsPayToPublicKeyHash())
        return 1;
    ...
```
(The same as in `CScript::Type()`)

We end up redoing some work, but it means that we can't possible pass in the wrong script type.

Reply:
Yes, I like both of your suggestions, will push another commit.